### DISCUSSION TITLE
Username-less (ID-less) WebAuthn UX authentication

### DISCUSSION THREAD

**Discussion by mabartos** [keycloak/keycloak#9821] [Ideas]
I'd like to discuss some options, how to improve usability of planned Username-less authentication regarding the UI. @vanrar68 created PR(#7860) for support it; thanks for that, I really like the idea and it'd be great to have it included in Keycloak. I quickly try the solution for that and it seems everything works as expected. However, I've encountered situation, where I think the UI can be better.

I created an authentication flow with `Username Password Form` and `WebAuthn Passwordless Authenticator` set up as `ALTERNATIVE` (for UsernamePwdForm I needed to include it to another flow, because it doesn't support `ALTERNATIVE` requirement). It means the `Username Password Form` is displayed to the user and it's possible to be authenticated only by the authenticator, which supports _Discoverable credential_. The thing is, it'd be great to improve UX within authenticator selection.

Implemented solution in the PR(#7860):
<img src="https://user-images.githubusercontent.com/38039883/151331848-0d31679c-4c04-484a-a6a8-8816ee48d4e6.png" width="350" />

As you can see, when there are some Identity providers defined, the section with the title `Or sign in with` is present and below the list of IdPs, there is another possibility(`Try another way`), how to sign in. It doesn't seem to me straightforward and IMHO the Username-less authentication should be on the same level like the Identity providers. And after clicking the `Try another way`, you're redirected to another page (which seems to me kinda unnecessary).
<img src="https://user-images.githubusercontent.com/38039883/151332661-60c44c2f-dca6-45b5-b5bd-5a6c3db92cd4.png" width="300" />

Proposed solution:
<img src="https://user-images.githubusercontent.com/38039883/151332837-b38acf51-38c6-4299-b98e-616538e04fef.png" width="350" />

As you can see, there's another item in the list with title `Security Key without Username`. I think, it'd be a better solution for that.
When there are defined more Identity providers and those items are restructured, the option for the Username-less could be the same like before. Or just edit the title to `Security Key` and keep the width same like for the IdPs?

<img src="https://user-images.githubusercontent.com/38039883/151333634-54328c35-ac91-4f60-99a8-d6834ea02300.png" width="350" />

@mposolda @vanrar68 @tnorimat WDYT?

EDIT: After discussion below, I agree to keep the design the same as in @vanrar68 approach except for the 'Try another way' link, which should be above the list of Identity Providers.

**Comment by vanrar68** [comment]
I'm so used to seeing identity providers shown as buttons that it's a bit strange to me to have "Security Key" look like one of them. In my mind things were organized like this:
- "Sign in with": other identity providers (remote)
- "Try Another Way": another auth method (local)

What would it look like with the following scenario:
ALT: UsernamePassword Form
ALT: UsernameForm + WebAuthnPasswordless
ALT: WebAuthnIDLess

We would end up with a login screen featuring:
- A UsernamePassword Form (local)
- A "Try another way" link allowing to switch to "UsernameForm" (local)
- A "Security Key" button (local) in the "Or Sign in with" area along with the identity providers (remote)

Anyway I agree the location of the "Try another way" link doesn't feel right when Identity Providers are configured.

Side note: Aren't you afraid such an implementation could break custom login forms (I guess it depends on the data model behind this implementation) ?
Side note 2: I'm definitely not a UI person, so I'll be OK with whatever decision is made on this topic

**Reply by mabartos** [reply]
@vanrar68 Thanks for your comment! I think you're right. It'd be kinda counterproductive regarding the usability and UX with another item `Security Key` among the IdPs with possibility to switch to a different authentication flow. Probably, it'd be sufficient only move the `Try another way` link above the list of IdPs.

**Comment by jbman** [comment]
A Webauthn Security Key is a replacement for username/password login. So I would also expect this option to be displayed right after the username password section. 
![image](https://user-images.githubusercontent.com/789496/152017865-8971eff5-6ad9-4c39-b973-821cc82f225c.png)
Maybe the section could have a header text like this:
```
Sign in to your account

<Username or email input>
<Password input>
[Login]
---
Passwordless
[Sign in with Security Key]
---
Or sign in with
[IdP 1]
[IdP 1]
```

It's much better to have it available on the top level screen, because at our Keycloak installations it is usually decided at realm level which methods or IdPs should be available to the user. Each of the sections could be available or not based on the realm settings

**Reply by Spechey** [reply]
From my point of view, I would keep @vanrar68's solution by default which has the merit of being generic, it's up to the administrator to make the evolution of the theme to have the UX that will be the most optimum to his need.

It is also necessary to think that it is perhaps the authentication X509 or another custom and not IdLess that one will want to see via the link "Try another way".

**Reply by mabartos** [reply]
Thanks for your comments. At this moment, it'd be good to keep the approach as @vanrar68 proposed except the `Try another way` link, which should be above the list of Identity Providers.

**Reply by vanrar68** [reply]
@mabartos how do you want to proceed ? should I update PR #7860 to include the new "Try another way" location or do you want to push your own UI PR ?

**Reply by mabartos** [reply]
@vanrar68 Feel free to do it. I can review it then.

**Reply by vanrar68** [reply]
Sorry @mabartos unfortunately I won't be able to push the UI update any time soon. I've rebased
@mposolda any objection to have the UI stuff in a dedicated PR ?

EDIT: I've rebased the PR against latest main branch. It includes an attempt at moving the "Try another way" link but have no idea if this is the right way to do. Let me know and I'll revert it if needed

**Reply by mabartos** [reply]
Thanks, @vanrar68! I'll take a look at it soon.
